File: app/src/main/java/com/xeniac/warrantyroster/mainactivity/warrantydetailsfragment/WarrantyDetailsFragment.java
package com.xeniac.warrantyroster.mainactivity.warrantydetailsfragment;

import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.ColorStateList;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.res.ResourcesCompat;
import androidx.core.widget.NestedScrollView;
import androidx.fragment.app.Fragment;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;

import com.apollographql.apollo.ApolloCall;
import com.apollographql.apollo.ApolloClient;
import com.apollographql.apollo.api.Response;
import com.apollographql.apollo.exception.ApolloException;
import com.apollographql.apollo.request.RequestHeaders;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.snackbar.BaseTransientBottomBar;
import com.google.android.material.snackbar.Snackbar;
import com.xeniac.warrantyroster.Constants;
import com.xeniac.warrantyroster.NetworkHelper;
import com.xeniac.warrantyroster.R;
import com.xeniac.warrantyroster.database.WarrantyRosterDatabase;
import com.xeniac.warrantyroster.databinding.FragmentWarrantyDetailsBinding;
import com.xeniac.warrantyroster.mainactivity.MainActivity;
import com.xeniac.warrantyroster.mainactivity.warrantiesfragment.WarrantyDataModel;

import java.text.DecimalFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Locale;
import java.util.Objects;

import ir.tapsell.plus.AdRequestCallback;
import ir.tapsell.plus.AdShowListener;
import ir.tapsell.plus.TapsellPlus;
import ir.tapsell.plus.TapsellPlusInitListener;
import ir.tapsell.plus.model.AdNetworkError;
import ir.tapsell.plus.model.AdNetworks;
import ir.tapsell.plus.model.TapsellPlusAdModel;
import ir.tapsell.plus.model.TapsellPlusErrorModel;

public class WarrantyDetailsFragment extends Fragment {

    private FragmentWarrantyDetailsBinding warrantyDetailsBinding;
    private View view;
    private Activity activity;
    private Context context;
    private NavController navController;
    private WarrantyRosterDatabase database;
    private WarrantyDataModel warranty;

    public WarrantyDetailsFragment() {
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        warrantyDetailsBinding = FragmentWarrantyDetailsBinding.inflate(inflater, container, false);
        view = warrantyDetailsBinding.getRoot();
        return view;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        WarrantyRosterDatabase.destroyInstance();
        warrantyDetailsBinding = null;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        activity = getActivity();
        context = getContext();
        navController = Navigation.findNavController(view);
        database = WarrantyRosterDatabase.getInstance(context);
        ((MainActivity) context).hideNavBar();

        returnToMainActivity();
        handleExtendedFAB();
        getWarranty();
        adInit();
        editWarrantyOnClick();
        deleteWarrantyOnClick();
    }

    private void returnToMainActivity() {
        warrantyDetailsBinding.toolbarWarrantyDetails.setNavigationOnClickListener(view ->
                activity.onBackPressed());
    }

    private void handleExtendedFAB() {
        warrantyDetailsBinding.svWarrantyDetails.setOnScrollChangeListener((NestedScrollView.OnScrollChangeListener)
                (v, scrollX, scrollY, oldScrollX, oldScrollY) -> {
                    if (scrollY > oldScrollY) {
                        warrantyDetailsBinding.fabWarrantyDetails.shrink();
                    } else {
                        warrantyDetailsBinding.fabWarrantyDetails.extend();
                    }
                });
    }

    private void getWarranty() {
        if (getArguments() != null) {
            WarrantyDetailsFragmentArgs args = WarrantyDetailsFragmentArgs.fromBundle(getArguments());
            warranty = args.getWarranty();
            long daysUntilExpiry = args.getDaysUntilExpiry();
            setWarrantyDetails(daysUntilExpiry);
        }
    }

    private void setWarrantyDetails(long daysUntilExpiry) {
        warrantyDetailsBinding.toolbarWarrantyDetails.setTitle(warranty.getTitle());
        warrantyDetailsBinding.tvWarrantyDetailsBrand.setText(warranty.getBrand());
        warrantyDetailsBinding.tvWarrantyDetailsModel.setText(warranty.getModel());
        warrantyDetailsBinding.tvWarrantyDetailsSerial.setText(warranty.getSerialNumber());
        warrantyDetailsBinding.tvWarrantyDetailsDescription.setText(warranty.getDescription());
        warrantyDetailsBinding.ivWarrantyDetailsIcon.setImageDrawable(ResourcesCompat.getDrawable(context.getResources(),
                database.categoryDAO().getCategoryById(warranty.getCategoryId()).getIcon(), context.getTheme()));

        Calendar startingCalendar = Calendar.getInstance();
        Calendar expiryCalendar = Calendar.getInstance();
        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-M-dd", Locale.getDefault());
        DecimalFormat decimalFormat = new DecimalFormat("00");

        try {
            startingCalendar.setTime(Objects.requireNonNull(dateFormat.parse(warranty.getStartingDate())));
            expiryCalendar.setTime(Objects.requireNonNull(dateFormat.parse(warranty.getExpiryDate())));

            warrantyDetailsBinding.tvWarrantyDetailsDateStarting.setText(String.format(("%s/%s/%s"),
                    decimalFormat.format(startingCalendar.get(Calendar.DAY_OF_MONTH)),
                    decimalFormat.format((startingCalendar.get(Calendar.MONTH)) + 1),
                    startingCalendar.get(Calendar.YEAR)));

            warrantyDetailsBinding.tvWarrantyDetailsDateExpiry.setText(String.format(("%s/%s/%s"),
                    decimalFormat.format(expiryCalendar.get(Calendar.DAY_OF_MONTH)),
                    decimalFormat.format((expiryCalendar.get(Calendar.MONTH)) + 1),
                    expiryCalendar.get(Calendar.YEAR)));

            if (daysUntilExpiry < 0) {
                warrantyDetailsBinding.tvWarrantyDetailsStatus.setText(context.getResources().getString(R.string.warranty_details_status_expired));
                warrantyDetailsBinding.tvWarrantyDetailsStatus.setTextColor(context.getResources().getColor(R.color.red));
                warrantyDetailsBinding.tvWarrantyDetailsStatus.setBackgroundTintList(ColorStateList.valueOf(context.getResources().getColor(R.color.red20)));
            } else if (daysUntilExpiry <= 30) {
                warrantyDetailsBinding.tvWarrantyDetailsStatus.setText(context.getResources().getString(R.string.warranty_details_status_soon));
                warrantyDetailsBinding.tvWarrantyDetailsStatus.setTextColor(context.getResources().getColor(R.color.orange));
                warrantyDetailsBinding.tvWarrantyDetailsStatus.setBackgroundTintList(ColorStateList.valueOf(context.getResources().getColor(R.color.orange20)));
            } else {
                warrantyDetailsBinding.tvWarrantyDetailsStatus.setText(context.getResources().getString(R.string.warranty_details_status_valid));
                warrantyDetailsBinding.tvWarrantyDetailsStatus.setTextColor(context.getResources().getColor(R.color.green));
                warrantyDetailsBinding.tvWarrantyDetailsStatus.setBackgroundTintList(ColorStateList.valueOf(context.getResources().getColor(R.color.green20)));
            }
        } catch (ParseException e) {
            e.printStackTrace();
            Log.e("setWarrantyDetails", "ParseException: " + e.getMessage());
        }
    }

    private void adInit() {
        TapsellPlus.initialize(context, Constants.TAPSELL_KEY, new TapsellPlusInitListener() {
            @Override
            public void onInitializeSuccess(AdNetworks adNetworks) {
                Log.i("adInit", "onInitializeSuccess: " + adNetworks.name());
            }

            @Override
            public void onInitializeFailed(AdNetworks adNetworks, AdNetworkError adNetworkError) {
                Log.e("adInit", "onInitializeFailed: " + adNetworks.name() + ", error: " + adNetworkError.getErrorMessage());
            }
        });
    }

    private void editWarrantyOnClick() {
        warrantyDetailsBinding.fabWarrantyDetails.setOnClickListener(view -> {
            WarrantyDetailsFragmentDirections.ActionWarrantyDetailsFragmentToEditWarrantyFragment action =
                    WarrantyDetailsFragmentDirections.actionWarrantyDetailsFragmentToEditWarrantyFragment(warranty);
            navController.navigate(action);
        });
    }

    private void deleteWarrantyOnClick() {
        warrantyDetailsBinding.toolbarWarrantyDetails.getMenu().getItem(0).setOnMenuItemClickListener(menuItem -> {
            deleteWarranty();
            return false;
        });
    }

    private void deleteWarranty() {
        if (NetworkHelper.hasNetworkAccess(context)) {
            showDeleteWarrantyDialog();
        } else {
            hideLoadingAnimation();
            Snackbar.make(view, context.getResources().getString(R.string.network_error_connection),
                    BaseTransientBottomBar.LENGTH_INDEFINITE)
                    .setAction(context.getResources().getString(R.string.network_error_retry), v -> deleteWarranty()).show();
        }
    }

    private void showDeleteWarrantyDialog() {
        MaterialAlertDialogBuilder deleteWarrantyDialogBuilder = new MaterialAlertDialogBuilder(context)
                .setTitle(context.getResources().getString(R.string.warranty_details_delete_dialog_title))
                .setMessage(String.format(context.getResources().getString(R.string.warranty_details_delete_dialog_message), warranty.getTitle()))
                .setPositiveButton(context.getResources().getString(R.string.warranty_details_delete_dialog_positive), (dialogInterface, i) ->
                        deleteWarrantyMutation())
                .setNegativeButton(context.getResources().getString(R.string.warranty_details_delete_dialog_negative), (dialogInterface, i) -> {
                });
        deleteWarrantyDialogBuilder.show();
    }

    private void deleteWarrantyMutation() {
        showLoadingAnimation();

        SharedPreferences loginPrefs = context.getSharedPreferences(Constants.PREFERENCE_LOGIN, Context.MODE_PRIVATE);
        String userToken = loginPrefs.getString(Constants.PREFERENCE_USER_TOKEN_KEY, null);

        ApolloClient apolloClient = ApolloClient.builder()
                .serverUrl(Constants.URL_GRAPHQL)
                .build();

        apolloClient.mutate(new DeleteWarrantyMutation(warranty.getId()))
                .toBuilder().requestHeaders(RequestHeaders.builder().addHeader("Authorization", "Bearer " + userToken).build())
                .build()
                .enqueue(new ApolloCall.Callback<DeleteWarrantyMutation.Data>() {
                    @Override
                    public void onResponse(@NonNull Response<DeleteWarrantyMutation.Data> response) {
                        if (warrantyDetailsBinding != null) {
                            activity.runOnUiThread(() -> {
                                hideLoadingAnimation();

                                if (!response.hasErrors()) {
                                    Log.i("deleteWarranty", "onResponse: " + response);
                                    Toast.makeText(context, String.format(
                                            context.getResources().getString(R.string.warranty_details_delete_success),
                                            warranty.getTitle()), Toast.LENGTH_LONG).show();
                                    requestInterstitialAd();
                                    activity.onBackPressed();
                                } else {
                                    Log.e("deleteWarranty", "onResponse Errors: " + response.getErrors());
                                    Snackbar.make(view, context.getResources().getString(R.string.network_error_response),
                                            BaseTransientBottomBar.LENGTH_INDEFINITE)
                                            .setAction(context.getResources().getString(R.string.network_error_retry), v -> deleteWarranty()).show();
                                }
                            });
                        }
                    }

                    @Override
                    public void onFailure(@NonNull ApolloException e) {
                        Log.e("deleteWarranty", "onFailure: " + e.getMessage());
                        if (warrantyDetailsBinding != null) {
                            activity.runOnUiThread(() -> {
                                hideLoadingAnimation();
                                Snackbar.make(view, context.getResources().getString(R.string.network_error_failure),
                                        BaseTransientBottomBar.LENGTH_LONG).show();
                            });
                        }
                    }
                });
    }

    private void showLoadingAnimation() {
        warrantyDetailsBinding.toolbarWarrantyDetails.getMenu().getItem(0).setVisible(false);
        warrantyDetailsBinding.fabWarrantyDetails.setClickable(false);
        warrantyDetailsBinding.cpiWarrantyDetails.setVisibility(View.VISIBLE);
    }

    private void hideLoadingAnimation() {
        warrantyDetailsBinding.toolbarWarrantyDetails.getMenu().getItem(0).setVisible(true);
        warrantyDetailsBinding.fabWarrantyDetails.setClickable(true);
        warrantyDetailsBinding.cpiWarrantyDetails.setVisibility(View.GONE);
    }

    private void requestInterstitialAd() {
        TapsellPlus.requestInterstitialAd(activity, Constants.DELETE_WARRANTY_Interstitial_ZONE_ID,
                new AdRequestCallback() {
                    @Override
                    public void response(TapsellPlusAdModel tapsellPlusAdModel) {
                        super.response(tapsellPlusAdModel);
                        Log.i("requestInterstitialAd", "response: " + tapsellPlusAdModel.toString());
                        showInterstitialAd(tapsellPlusAdModel.getResponseId());
                    }

                    @Override
                    public void error(String s) {
                        super.error(s);
                        Log.e("requestInterstitialAd", "error: " + s);
                    }
                });
    }

    private void showInterstitialAd(String responseId) {
        TapsellPlus.showInterstitialAd(activity, responseId, new AdShowListener() {
            @Override
            public void onOpened(TapsellPlusAdModel tapsellPlusAdModel) {
                super.onOpened(tapsellPlusAdModel);
            }

            @Override
            public void onClosed(TapsellPlusAdModel tapsellPlusAdModel) {
                super.onClosed(tapsellPlusAdModel);
            }

            @Override
            public void onError(TapsellPlusErrorModel tapsellPlusErrorModel) {
                super.onError(tapsellPlusErrorModel);
            }
        });
    }
}

File: app/src/main/java/com/xeniac/warrantyroster_manager/mainactivity/warrantydetailsfragment/WarrantyDetailsFragment.kt
package com.xeniac.warrantyroster_manager.mainactivity.warrantydetailsfragment

import android.os.Bundle
import android.util.Log
import android.view.Gravity
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.widget.Toast
import androidx.core.content.ContextCompat
import androidx.core.widget.NestedScrollView
import androidx.fragment.app.Fragment
import androidx.navigation.NavController
import androidx.navigation.Navigation
import androidx.navigation.fragment.navArgs
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.snackbar.BaseTransientBottomBar.LENGTH_INDEFINITE
import com.google.android.material.snackbar.BaseTransientBottomBar.LENGTH_LONG
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.firestore.ktx.firestore
import com.google.firebase.ktx.Firebase
import com.xeniac.warrantyroster_manager.Constants
import com.xeniac.warrantyroster_manager.NetworkHelper
import com.xeniac.warrantyroster_manager.R
import com.xeniac.warrantyroster_manager.database.WarrantyRosterDatabase
import com.xeniac.warrantyroster_manager.databinding.FragmentWarrantyDetailsBinding
import com.xeniac.warrantyroster_manager.mainactivity.MainActivity
import com.xeniac.warrantyroster_manager.model.Warranty
import ir.tapsell.plus.TapsellPlus
import ir.tapsell.plus.TapsellPlusInitListener
import ir.tapsell.plus.model.AdNetworkError
import ir.tapsell.plus.model.AdNetworks
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.withContext
import java.text.DecimalFormat
import java.text.SimpleDateFormat
import java.util.*

class WarrantyDetailsFragment : Fragment(R.layout.fragment_warranty_details) {

    private var _binding: FragmentWarrantyDetailsBinding? = null
    private val binding get() = _binding!!
    private lateinit var navController: NavController
    private lateinit var database: WarrantyRosterDatabase
    private lateinit var warranty: Warranty
    private val warrantiesCollectionRef =
        Firebase.firestore.collection(Constants.COLLECTION_WARRANTIES)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        _binding = FragmentWarrantyDetailsBinding.bind(view)
        navController = Navigation.findNavController(view)
        database = WarrantyRosterDatabase.getInstance(requireContext())
        (requireContext() as MainActivity).hideNavBar()

        returnToMainActivity()
        handleExtendedFAB()
        getWarranty()
        adInit()
        editWarrantyOnClick()
        deleteWarrantyOnClick()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        WarrantyRosterDatabase.destroyInstance()
        _binding = null
    }

    private fun returnToMainActivity() {
        binding.toolbarWarrantyDetails.setNavigationOnClickListener {
            requireActivity().onBackPressed()
        }
    }

    private fun handleExtendedFAB() {
        binding.svWarrantyDetails.setOnScrollChangeListener(
            NestedScrollView.OnScrollChangeListener { _, _, scrollY, _, oldScrollY ->
                if (scrollY > oldScrollY) {
                    binding.fabWarrantyDetails.shrink()
                } else {
                    binding.fabWarrantyDetails.extend()
                }
            })
    }

    private fun getWarranty() {
        val args: WarrantyDetailsFragmentArgs by navArgs()
        warranty = args.warranty
        val daysUntilExpiry = args.daysUntilExpiry
        setWarrantyDetails(daysUntilExpiry)
    }

    private fun setWarrantyDetails(daysUntilExpiry: Long) {
        if (warranty.brand != null) {
            binding.tvWarrantyDetailsBrand.text = warranty.brand
        } else {
            binding.tvWarrantyDetailsBrand.setTextColor(
                ContextCompat.getColor(requireContext(), R.color.grayDark)
            )
            binding.tvWarrantyDetailsBrand.text =
                requireContext().getString(R.string.warranty_details_empty_device)
        }

        if (warranty.model != null) {
            binding.tvWarrantyDetailsModel.text = warranty.model
        } else {
            binding.tvWarrantyDetailsModel.setTextColor(
                ContextCompat.getColor(requireContext(), R.color.grayDark)
            )
            binding.tvWarrantyDetailsModel.text =
                requireContext().getString(R.string.warranty_details_empty_device)
        }

        if (warranty.serialNumber != null) {
            binding.tvWarrantyDetailsSerial.text = warranty.serialNumber
        } else {
            binding.tvWarrantyDetailsSerial.setTextColor(
                ContextCompat.getColor(requireContext(), R.color.grayDark)
            )
            binding.tvWarrantyDetailsSerial.text =
                requireContext().getString(R.string.warranty_details_empty_device)
        }

        if (warranty.description != null) {
            binding.tvWarrantyDetailsDescription.text = warranty.description
        } else {
            binding.tvWarrantyDetailsDescription.gravity = Gravity.CENTER
            binding.tvWarrantyDetailsDescription.setTextColor(
                ContextCompat.getColor(requireContext(), R.color.grayDark)
            )
            binding.tvWarrantyDetailsDescription.text =
                requireContext().getString(R.string.warranty_details_empty_description)
        }

        binding.toolbarWarrantyDetails.title = warranty.title
        binding.ivWarrantyDetailsIcon.setImageResource(
            database.categoryDAO().getCategoryById(warranty.categoryId).icon
        )

        val startingCalendar = Calendar.getInstance()
        val expiryCalendar = Calendar.getInstance()
        val dateFormat = SimpleDateFormat("yyyy-M-dd")
        val decimalFormat = DecimalFormat("00")

        dateFormat.parse(warranty.startingDate)?.let {
            startingCalendar.time = it
        }

        dateFormat.parse(warranty.expiryDate)?.let {
            expiryCalendar.time = it
        }

        binding.tvWarrantyDetailsDateStarting.text =
            "${decimalFormat.format((startingCalendar.get(Calendar.MONTH)) + 1)}/" +
                    "${decimalFormat.format(startingCalendar.get(Calendar.DAY_OF_MONTH))}/" +
                    "${startingCalendar.get(Calendar.YEAR)}"

        binding.tvWarrantyDetailsDateExpiry.text =
            "${decimalFormat.format((expiryCalendar.get(Calendar.MONTH)) + 1)}/" +
                    "${decimalFormat.format(expiryCalendar.get(Calendar.DAY_OF_MONTH))}/" +
                    "${expiryCalendar.get(Calendar.YEAR)}"

        when {
            daysUntilExpiry < 0 -> {
                binding.tvWarrantyDetailsStatus.text =
                    requireContext().getString(R.string.warranty_details_status_expired)
                binding.tvWarrantyDetailsStatus.setTextColor(
                    ContextCompat.getColor(requireContext(), R.color.red)
                )
                binding.tvWarrantyDetailsStatus.backgroundTintList =
                    ContextCompat.getColorStateList(requireContext(), R.color.red20)
            }
            daysUntilExpiry <= 30 -> {
                binding.tvWarrantyDetailsStatus.text =
                    requireContext().getString(R.string.warranty_details_status_soon)
                binding.tvWarrantyDetailsStatus.setTextColor(
                    ContextCompat.getColor(requireContext(), R.color.orange)
                )
                binding.tvWarrantyDetailsStatus.backgroundTintList =
                    ContextCompat.getColorStateList(requireContext(), R.color.orange20)
            }
            else -> {
                binding.tvWarrantyDetailsStatus.text =
                    requireContext().getString(R.string.warranty_details_status_valid)
                binding.tvWarrantyDetailsStatus.setTextColor(
                    ContextCompat.getColor(requireContext(), R.color.green)
                )
                binding.tvWarrantyDetailsStatus.backgroundTintList =
                    ContextCompat.getColorStateList(requireContext(), R.color.green20)
            }
        }
    }

    private fun adInit() {
        TapsellPlus.initialize(
            requireContext(), Constants.TAPSELL_KEY, object : TapsellPlusInitListener {
                override fun onInitializeSuccess(adNetworks: AdNetworks?) {
                    Log.i("adInit", "onInitializeSuccess: ${adNetworks?.name}")
                }

                override fun onInitializeFailed(
                    adNetworks: AdNetworks?, adNetworkError: AdNetworkError?
                ) {
                    Log.e(
                        "adInit",
                        "onInitializeFailed: ${adNetworks?.name}, error: ${adNetworkError?.errorMessage}"
                    )
                }
            })
    }

    private fun editWarrantyOnClick() {
        binding.fabWarrantyDetails.setOnClickListener {
            val action = WarrantyDetailsFragmentDirections
                .actionWarrantyDetailsFragmentToEditWarrantyFragment(warranty)
            navController.navigate(action)
        }
    }

    private fun deleteWarrantyOnClick() {
        binding.toolbarWarrantyDetails.menu.getItem(0).setOnMenuItemClickListener {
            deleteWarranty()
            false
        }
    }

    private fun deleteWarranty() {
        if (NetworkHelper.hasNetworkAccess(requireContext())) {
            showDeleteWarrantyDialog()
        } else {
            hideLoadingAnimation()
            Snackbar.make(
                binding.root,
                requireContext().getString(R.string.network_error_connection),
                LENGTH_INDEFINITE
            ).apply {
                setAction(requireContext().getString(R.string.network_error_retry)) { deleteWarranty() }
                show()
            }
        }
    }

    private fun showDeleteWarrantyDialog() {
        MaterialAlertDialogBuilder(requireContext()).apply {
            setTitle(requireContext().getString(R.string.warranty_details_delete_dialog_title))
            setMessage(
                String.format(
                    requireContext().getString(R.string.warranty_details_delete_dialog_message),
                    warranty.title
                )
            )
            setPositiveButton(requireContext().getString(R.string.warranty_details_delete_dialog_positive)) { _, _ -> deleteWarrantyFromFirestore() }
            setNegativeButton(requireContext().getText(R.string.warranty_details_delete_dialog_negative)) { _, _ -> }
            show()
        }
    }

    private fun deleteWarrantyFromFirestore() {
        showLoadingAnimation()
        CoroutineScope(Dispatchers.IO).launch {
            try {
                warrantiesCollectionRef.document(warranty.id).delete().await()
                Log.i("deleteWarranty", "${warranty.id} successfully deleted.")
                withContext(Dispatchers.Main) {
                    hideLoadingAnimation()
                    Toast.makeText(
                        requireContext(),
                        String.format(
                            requireContext().getString(
                                R.string.warranty_details_delete_success,
                                warranty.title
                            )
                        ),
                        Toast.LENGTH_LONG
                    ).show()
                    (requireContext() as MainActivity).requestInterstitialAd()
                    requireActivity().onBackPressed()
                }
            } catch (e: Exception) {
                Log.e("deleteWarranty", "Exception: ${e.message}")
                withContext(Dispatchers.Main) {
                    hideLoadingAnimation()
                    Snackbar.make(
                        binding.root,
                        requireContext().getString(R.string.network_error_failure),
                        LENGTH_LONG
                    ).show()
                }
            }
        }
    }

    private fun showLoadingAnimation() {
        binding.toolbarWarrantyDetails.menu.getItem(0).isVisible = false
        binding.fabWarrantyDetails.isClickable = false
        binding.cpiWarrantyDetails.visibility = VISIBLE
    }

    private fun hideLoadingAnimation() {
        binding.toolbarWarrantyDetails.menu.getItem(0).isVisible = true
        binding.fabWarrantyDetails.isClickable = true
        binding.cpiWarrantyDetails.visibility = GONE
    }
}